### PR TITLE
Fix SpriteEffect not updating ScreenMap

### LINKED ISSUE #14047: Game crashes if a SpriteEffect starts with an empty frame
Regression from #13979.

The problem will be the fact that compressed shp(ts) files with empty first frame return an `Animation.Image` size of `0, 0`.

One possible solution would be to check for the largest frame of the animation and return that as `Animation.LargestImage` so we can pass that to the ScreenMap to calculate the bounds, but I'm not sure what performance impact this approach might have.

### PR DESCRIPTION
**Pull request by reaperrr**
Fixes #14047.

As additional safe-guarding against zero-bounds, the adding/removing in ScreenMap was tweaked as well, now it only adds effects if W/H are not 0.

### CODE REVIEW

**Comment by reaperrr** [comment]
Updated.

File: OpenRA.Game/Traits/World/ScreenMap.cs [hunk 1]
```diff
@@ -117,9 +120,14 @@ public void Update(IEffect effect, WPos position, Sprite sprite)
...
-		public void Remove(IEffect e)
+		public void Remove(IEffect effect)
 		{
-			partitionedEffects.Remove(e);
+			partitionedEffects.Remove(effect);
+		}
+
+		bool ValidBounds(Rectangle bounds)
+		{
+			return bounds.Width != 0 && bounds.Height != 0;
```
Comment (Mailaender) [inline OpenRA.Game/Traits/World/ScreenMap.cs]: Can be simplified to `bounds != Rectangle.Empty` or maybe changed to `bounds.Width > 0 && bounds.Height > 0`.

**Comment by RoosterDragon** [comment]
Why are their effects with bounds of zero in the first place? What use do they have? I know we had to exclude some actors with zero bounds because they're invisible markers for things, is that the case here as well?

If we do need the bounds check, the way it was applied previously for actors and such was further up the call stack:
```c#
if (!self.Bounds.Size.IsEmpty)
    self.World.ScreenMap.Add(self);
```

So I guess we would want similar checks for consistency, if we do need to add them.

**Comment by pchote** [comment]
For actors we have always used a fixed size regardless of their current rendered state, but for effects we are updating them every tick to reflect what is actually visible.  The issue comes when the effect legitimately doesn't want to render something during one of the frames of its animation.

IMO handing this lower level in the stack is cleaner, and we will eventually need to do something similar for actors as part of the fix for bibs and shadows.

**Comment by reaperrr** [comment]
Updated.

**Review by Mailaender — APPROVED** [review]

**Review by pchote — APPROVED** [review]
I don't have a repro case for the original issue so I can't verify that this is sufficient to fix the crash, but these will definitely help.

**Comment by pchote** [comment]
https://github.com/OpenRA/OpenRA/wiki/Changelog-(bleed)/_compare/9520a6ba0009045d27939d4b2a1350f839a04ab7

**Pull request merged**